### PR TITLE
In memory support

### PR DESCRIPTION
**Pull request by idohalevi**

### CODE REVIEW

File: src/main/java/io/logz/log4j2/LogzioAppender.java [hunk 1]
```diff
@@ -263,37 +295,80 @@ public void start() {
...
         }
-        String bufferDirPath;
-        if (bufferDir != null) {
-            bufferDirPath = bufferDir;
-            File bufferFile = new File(bufferDirPath);
-            if (bufferFile.exists()) {
-                if (!bufferFile.canWrite()) {
-                    statusLogger.error("We cant write to your bufferDir location: "+bufferFile.getAbsolutePath());
+
+        if (!inMemoryBuffer) {
```
Comment (segalziv) [inline src/main/java/io/logz/log4j2/LogzioAppender.java]: extract logic to methods for readability

File: src/main/java/io/logz/log4j2/LogzioAppender.java [hunk 2]
```diff
@@ -263,37 +295,80 @@ public void start() {
...
+            if (!(fileSystemFullPercentThreshold >= 1 && fileSystemFullPercentThreshold <= 100)) {
```
Comment (segalziv) [inline src/main/java/io/logz/log4j2/LogzioAppender.java]: extract to method

File: src/main/java/io/logz/log4j2/LogzioAppender.java [hunk 3]
```diff
@@ -263,37 +295,80 @@ public void start() {
...
+                    }
+                }
+            }
+            else {
+                bufferDirPath = System.getProperty("java.io.tmpdir") + File.separator+"logzio-log4j2-buffer";
+            }
+            File bufferDirFile = new File(bufferDirPath, logzioType);
+            tasksExecutor = Executors.newScheduledThreadPool(3, Log4jThreadFactory.createDaemonThreadFactory(this.getClass().getSimpleName()));
+            try {
+                logzioSender = LogzioSender
```
Comment (segalziv) [inline src/main/java/io/logz/log4j2/LogzioAppender.java]: looking good!

File: src/main/java/io/logz/log4j2/LogzioAppender.java [hunk 4]
```diff
@@ -263,37 +295,80 @@ public void start() {
...
+        }else {
+            if (bufferSizeThreshold <= 0) {
+                if (bufferSizeThreshold != -1) {
+                    statusLogger.error("bufferSizeThreshold should be a non zero integer, or -1");
                     return;
                 }
             }
+            tasksExecutor = Executors.newScheduledThreadPool(1, Log4jThreadFactory.createDaemonThreadFactory(this.getClass().getSimpleName()));
+            try {
+                logzioSender = LogzioSender
```
Comment (segalziv) [inline src/main/java/io/logz/log4j2/LogzioAppender.java]: lets join the builders in the if-else to one builder and build() in the end

File: src/test/java/io/logz/log4j2/InMemoryBuffer.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
+package io.logz.log4j2;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class InMemoryBuffer extends Log4j2AppenderTest {
```
Comment (segalziv) [inline src/test/java/io/logz/log4j2/InMemoryBuffer.java]: InMemoryBuffer  -> InMemoryBufferTest

File: src/test/java/io/logz/log4j2/Log4j2AppenderTest.java [hunk 6]
```diff
@@ -212,36 +213,8 @@ public void testTokenAndLogzioUrlFromSystemEnvironment() {
...
-
-        LogzioAppender appender = logzioLog4j2AppenderBuilder.build();
-        appender.start();
-        assertThat(appender.isStarted()).isTrue();
-        ((org.apache.logging.log4j.core.Logger) log4j2Logger).addAppender(appender);
-        ((org.apache.logging.log4j.core.Logger) log4j2Logger).setAdditive(false);
-
-        return log4j2Logger;
-    }
+    abstract Logger createLogger(String token, String type, String loggerName, Integer drainTimeout,
```
Comment (segalziv) [inline src/test/java/io/logz/log4j2/Log4j2AppenderTest.java]: lets use parametrized list - https://github.com/junit-team/junit4/wiki/parameterized-tests
and remove the extends

File: src/main/java/io/logz/log4j2/LogzioAppender.java [hunk 7]
```diff
@@ -263,37 +295,80 @@ public void start() {
...
+                          .setFsPercentThreshold(fileSystemFullPercentThreshold)
+                          .setGcPersistedQueueFilesIntervalSeconds(gcPersistedQueueFilesIntervalSeconds)
+                        .EndDiskQueue()
+                        .build();
+            } catch (LogzioParameterErrorException e) {
+                statusLogger.error("Some of the configuration parameters of logz.io is wrong: "+e.getMessage(),e);
+                return;
+            }
+        }else {
+            if (bufferSizeThreshold <= 0) {
```
Comment (amirkalron) [inline src/main/java/io/logz/log4j2/LogzioAppender.java]: why not:
if(bufferSizeThreshold <= 0 && bufferSizeThreshold != -1){
                    statusLogger.error("bufferSizeThreshold should be a non zero integer, or -1");
                    return;
}

File: src/main/java/io/logz/log4j2/LogzioAppender.java [hunk 8]
```diff
@@ -108,10 +109,18 @@ public static Builder newBuilder() {
         @PluginBuilderAttribute
         private boolean compressRequests = false;
 
+        @PluginBuilderAttribute
+        private boolean inMemoryBuffer = false;
```
Comment (asafm) [inline src/main/java/io/logz/log4j2/LogzioAppender.java]: Let's use queue instead of buffer

File: src/main/java/io/logz/log4j2/LogzioAppender.java [hunk 9]
```diff
@@ -108,10 +109,18 @@ public static Builder newBuilder() {
...
+
+        @PluginBuilderAttribute
+        private int bufferSizeThreshold  = 100 * 1024 *1024;
```
Comment (asafm) [inline src/main/java/io/logz/log4j2/LogzioAppender.java]: add units to name of variable, and use queue instead of buffer.

File: src/main/java/io/logz/log4j2/LogzioAppender.java [hunk 10]
```diff
@@ -210,15 +228,16 @@ public Builder setCompressRequests(boolean compressRequests) {
     private final boolean addHostname;
     private final int gcPersistedQueueFilesIntervalSeconds;
     private final boolean compressRequests;
-
+    private final boolean inMemoryBuffer;
+    private final int bufferSizeThreshold;
```
Comment (asafm) [inline src/main/java/io/logz/log4j2/LogzioAppender.java]: long

File: src/main/java/io/logz/log4j2/LogzioAppender.java (hunk 1 above)
Comment (asafm) [inline src/main/java/io/logz/log4j2/LogzioAppender.java]: Use this flag, in the middle of the builder, only for the section you're building the queue, thus save boilerplate code.

File: src/main/java/io/logz/log4j2/LogzioAppender.java [hunk 11]
```diff
@@ -348,7 +423,7 @@ private JsonObject formatMessageAsJson(LogEvent loggingEvent) {
 
     private static String getValueFromSystemEnvironmentIfNeeded(String value) {
         if (value == null)
-            return value;
+            return null;
```
Comment (asafm) [inline src/main/java/io/logz/log4j2/LogzioAppender.java]: Why this change?

File: src/test/java/io/logz/log4j2/Log4j2AppenderTest.java (hunk 6 above)
Comment (asafm) [reply src/test/java/io/logz/log4j2/Log4j2AppenderTest.java]: But he creates a different logger

File: src/test/java/io/logz/log4j2/OnDiskBuffer.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
+package io.logz.log4j2;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class OnDiskBuffer extends Log4j2AppenderTest {
```
Comment (asafm) [inline src/test/java/io/logz/log4j2/OnDiskBuffer.java]: append "Test" to class name

File: src/main/java/io/logz/log4j2/LogzioAppender.java (hunk 11 above)
Comment (idohalevi) [reply src/main/java/io/logz/log4j2/LogzioAppender.java]: @asafm I think it's clearer what is the return value

File: src/test/java/io/logz/log4j2/OnDiskBuffer.java (hunk 12 above)
Comment (idohalevi) [reply src/test/java/io/logz/log4j2/OnDiskBuffer.java]: @asafm Going to use Parameterized instead

File: src/main/java/io/logz/log4j2/LogzioAppender.java (hunk 11 above)
Comment (asafm) [reply src/main/java/io/logz/log4j2/LogzioAppender.java]: right

File: src/test/java/io/logz/log4j2/OnDiskBuffer.java (hunk 12 above)
Comment (asafm) [reply src/test/java/io/logz/log4j2/OnDiskBuffer.java]: Just make sure it's really easier to read in that form as I found it hard to understand.

File: src/main/java/io/logz/log4j2/LogzioAppender.java [hunk 13]
```diff
@@ -81,9 +79,13 @@ public static Builder newBuilder() {
         @PluginBuilderAttribute
         int fileSystemFullPercentThreshold = 98;
 
+        @Deprecated
```
Comment (segalziv) [inline src/main/java/io/logz/log4j2/LogzioAppender.java]: I prefer that when adding @Deprecated to state what is the alternative, or why

File: src/main/java/io/logz/log4j2/LogzioAppender.java [hunk 14]
```diff
@@ -249,12 +285,50 @@ private LogzioAppender(String name, Filter filter, final boolean ignoreException
     }
 
     public void start() {
-        if (!(fileSystemFullPercentThreshold >= 1 && fileSystemFullPercentThreshold <= 100)) {
-            if (fileSystemFullPercentThreshold != -1) {
-                statusLogger.error("fileSystemFullPercentThreshold should be a number between 1 and 100, or -1");
-                return;
-            }
+        HttpsRequestConfiguration conf = getHttpsRequestConfiguration();
+        if (conf == null) return;
```
Comment (segalziv) [inline src/main/java/io/logz/log4j2/LogzioAppender.java]: lets add a log on this case stating that the conf was missing?

File: README.md [hunk 15]
```diff
@@ -52,6 +52,7 @@ This appender uses  [LogzioSender](https://github.com/logzio/logzio-java-sender)
 | Parameter          | Default                              | Explained  |
 | ------------------ | ------------------------------------ | ----- |
 | **inMemoryQueueCapacityBytes**       | *1024 * 1024 * 100*                                | The amount of memory(bytes) we are allowed to use for the memory queue. If the value is -1 the sender will not limit the queue size.|
+| **inMemoryLogsCountLimit**       | *-1*                                | Number of logs we are allowed to have in the queue before dropping logs. If the value is -1 the sender will not limit the number of logs allowed.|
```
Comment (asafm) [inline README.md]: Limit of bytes you call it capacity - `inMemoryQueueCapacityBytes`
Limit of logs you call it limit. 
I suggest using capacity as well

File: src/main/java/io/logz/log4j2/LogzioAppender.java [hunk 16]
```diff
@@ -111,20 +111,23 @@ public static Builder newBuilder() {
...
         @PluginBuilderAttribute
-        private boolean inMemoryQueue = false;
+        boolean inMemoryQueue = false;
 
         @PluginBuilderAttribute
-        private long inMemoryQueueCapacityBytes  = 100 * 1024 *1024;
+        long inMemoryQueueCapacityBytes  = 100 * 1024 *1024;
+
+        @PluginBuilderAttribute
+        long inMemoryLogsCountLimit  = DONT_LIMIT_QUEUE_SPACE;
```
Comment (asafm) [inline src/main/java/io/logz/log4j2/LogzioAppender.java]: DONT_LIMIT_QUEUE_SPACE - this name is odd. You're limiting space?

File: src/main/java/io/logz/log4j2/LogzioAppender.java [hunk 17]
```diff
@@ -370,6 +381,10 @@ private boolean validateQueueCapacity() {
             statusLogger.error("inMemoryQueueCapacityBytes should be a non zero integer or -1");
             return false;
         }
+        if (inMemoryLogsCountLimit <= 0 && inMemoryLogsCountLimit != DONT_LIMIT_QUEUE_SPACE) {
```
Comment (asafm) [inline src/main/java/io/logz/log4j2/LogzioAppender.java]: `&&` should be `||` ?

File: src/main/java/io/logz/log4j2/LogzioAppender.java [hunk 18]
```diff
@@ -370,6 +381,10 @@ private boolean validateQueueCapacity() {
...
+            statusLogger.error("inMemoryLogsCountLimit should be a non zero integer or -1");
```
Comment (asafm) [inline src/main/java/io/logz/log4j2/LogzioAppender.java]: or" + DONT_LIMIT_QUEUE_SPACE

File: src/test/java/io/logz/log4j2/InMemoryTest.java [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+public class InMemoryTest extends BaseLog4jAppenderTest {
+    private Builder logzioAppenderBuilder;
+
+    @Before
+    public void setNewLogzioAppenderBuilder() {
+        logzioAppenderBuilder = new LogzioAppender.Builder().setInMemoryQueue(true);
+    }
+
+    @Test
+    public void validateQueueCapacityLimit() {
```
Comment (asafm) [inline src/test/java/io/logz/log4j2/InMemoryTest.java]: Capacity and Limit as synonyms

File: src/test/java/io/logz/log4j2/Log4j2AppenderTest.java [hunk 20]
```diff
@@ -18,47 +14,36 @@
...
-    public void startMockListener() throws Exception {
-        mockListener = new io.logz.test.MockLogzioBulkListener();
-        mockListener.start();
-    }
-
-    @After
-    public void stopMockListener() {
-        mockListener.stop();
+    public void setNewLogzioAppenderBuilder() {
+        logzioAppenderBuilder = queueType == QueueType.DISK ?
```
Comment (asafm) [inline src/test/java/io/logz/log4j2/Log4j2AppenderTest.java]: Syntax is:
```
logzioAppenderBuilder = queueType == QueueType.DISK 
              ? new LogzioAppender.Builder() 
              :  new LogzioAppender.Builder().setInMemoryQueue(true);
```

File: src/test/java/io/logz/log4j2/Log4j2AppenderTest.java [hunk 21]
```diff
@@ -18,47 +14,36 @@
...
     @Parameterized.Parameters
-    public static Collection<Object[]> logzioSenderBuilders() {
-        List<Builder> builders = new ArrayList<>();
-        builders.add(new Builder());
-        builders.add(new Builder().setInMemoryQueue(true));
-
-        Collection<Object[]> result = new ArrayList<>();
-        for (Builder builder : builders) {
-            result.add(new Object[]{builder});
+    public static Collection<QueueType[]> logzioSenderBuilders() {
```
Comment (asafm) [inline src/test/java/io/logz/log4j2/Log4j2AppenderTest.java]: the method name doesn't correlate with the type (QueueType)

File: src/main/java/io/logz/log4j2/LogzioAppender.java (hunk 17 above)
Comment (idohalevi) [reply src/main/java/io/logz/log4j2/LogzioAppender.java]: @asafm The only way for `inMemoryLogsCountLimit ` to be negative and valid is if it is equal to -1
